### ISSUE TITLE
AutoExpand in sub child

### ISSUE THREAD

**Issue by agminformatique** [OData/WebApi#1145]
I have a problem with AutoExpand attribute when it's used in a sub child.
I need to expand a child in an AutoExpand entity.

Assemblies affected

Microsoft.AspNet.OData v6.1.0

Reproduce steps

I use 4 entites
class Entity1
{
int identite;
string libelle;
EntityType entityType;
}
class EntityType
{
int idEntityType;
string libelle;
EntityAutoExpand entityAutoExpand;
}
[AutoExpand]
class EntityAutoExpand
{
int idEntityAutoExpand;
string libelle;
List listSubItems;
}
class EntityAutoExpandSubItems
{
int idEntityAutoExpandSubItems;
string libelle;
}

When I call my WebService on EntityType with
https://localhost:9443/odata/EntityType?$expand=entityAutoExpand($expand=listSubItems)
Or
https://localhost:9443/odata/EntityType?$expand=entityAutoExpand
that's work would expected.

But if I call on Entity1
https://localhost:9443/odata/Entity1?&$expand=EntityType($expand=EntityAutoExpand($expand=EntityAutoExpandSubItems))
Or
https://localhost:9443/odata/Entity1?&$expand=EntityType($expand=EntityAutoExpand)
I have an error 500 with message "An element with the same key have already be added"

Additional detail

When I remove AutoExpand attribute, that's work.

Thanks in advance

**Comment by xuzhg** [comment]
@agminformatique 

Thanks your input. It's indeed a bug in Web API OData Auto Expand feature.
The problem is located at https://github.com/OData/WebApi/blob/master/src/System.Web.OData/OData/Query/SelectExpandQueryOption.cs#L543-L545

We will triage it and figure out a release to fix it. 

Thanks,
-ODataTeam

**Comment by agminformatique** [comment]
Hello
That's a good news
Thanks

**Comment by xuzhg** [comment]
I'd like to leave it open until it's fixed. Is it ok?

**Comment by agminformatique** [comment]
Sorry of course

**Comment by davidmilligan** [comment]
I'm having what I think is the same issue. If I set a property to AutoExpand, and then also include that property in the $expand in the request, Web API OData throws an exception:

```
System.ArgumentException: An item with the same key has already been added.
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at System.Web.OData.Formatter.Serialization.SelectExpandNode.BuildExpansions(SelectExpandClause selectExpandClause, HashSet`1 allNavigationProperties)
   at System.Web.OData.Formatter.Serialization.SelectExpandNode..ctor(SelectExpandClause selectExpandClause, IEdmStructuredType structuredType, IEdmModel model)
   at System.Web.OData.Formatter.Serialization.ODataResourceSerializer.CreateSelectExpandNode(ResourceContext resourceContext)
   at System.Web.OData.Formatter.Serialization.ODataResourceSerializer.WriteResource(Object graph, ODataWriter writer, ODataSerializerContext writeContext, IEdmTypeReference expectedType)
   at System.Web.OData.Formatter.Serialization.ODataResourceSerializer.WriteObjectInline(Object graph, IEdmTypeReference expectedType, ODataWriter writer, ODataSerializerContext writeContext)
   at System.Web.OData.Formatter.Serialization.ODataResourceSetSerializer.WriteResourceSet(IEnumerable enumerable, IEdmTypeReference resourceSetType, ODataWriter writer, ODataSerializerContext writeContext)
   at System.Web.OData.Formatter.Serialization.ODataResourceSetSerializer.WriteObjectInline(Object graph, IEdmTypeReference expectedType, ODataWriter writer, ODataSerializerContext writeContext)
   at System.Web.OData.Formatter.Serialization.ODataResourceSerializer.WriteComplexAndExpandedNavigationProperty(IEdmProperty edmProperty, SelectExpandClause selectExpandClause, ResourceContext resourceContext, ODataWriter writer)
   at System.Web.OData.Formatter.Serialization.ODataResourceSerializer.WriteExpandedNavigationProperties(IDictionary`2 navigationPropertiesToExpand, ResourceContext resourceContext, ODataWriter writer)
   at System.Web.OData.Formatter.Serialization.ODataResourceSerializer.WriteResource(Object graph, ODataWriter writer, ODataSerializerContext writeContext, IEdmTypeReference expectedType)
   at System.Web.OData.Formatter.Serialization.ODataResourceSerializer.WriteObjectInline(Object graph, IEdmTypeReference expectedType, ODataWriter writer, ODataSerializerContext writeContext)
   at System.Web.OData.Formatter.Serialization.ODataResourceSerializer.WriteComplexAndExpandedNavigationProperty(IEdmProperty edmProperty, SelectExpandClause selectExpandClause, ResourceContext resourceContext, ODataWriter writer)
   at System.Web.OData.Formatter.Serialization.ODataResourceSerializer.WriteExpandedNavigationProperties(IDictionary`2 navigationPropertiesToExpand, ResourceContext resourceContext, ODataWriter writer)
   at System.Web.OData.Formatter.Serialization.ODataResourceSerializer.WriteResource(Object graph, ODataWriter writer, ODataSerializerContext writeContext, IEdmTypeReference expectedType)
   at System.Web.OData.Formatter.Serialization.ODataResourceSerializer.WriteObjectInline(Object graph, IEdmTypeReference expectedType, ODataWriter writer, ODataSerializerContext writeContext)
   at System.Web.OData.Formatter.Serialization.ODataResourceSerializer.WriteObject(Object graph, Type type, ODataMessageWriter messageWriter, ODataSerializerContext writeContext)
   at System.Web.OData.Formatter.ODataMediaTypeFormatter.WriteToStream(Type type, Object value, Stream writeStream, HttpContent content, HttpContentHeaders contentHeaders)
   at System.Web.OData.Formatter.ODataMediaTypeFormatter.WriteToStreamAsync(Type type, Object value, Stream writeStream, HttpContent content, TransportContext transportContext, CancellationToken cancellationToken)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.OData.Batch.ODataBatchResponseItem.<WriteMessageAsync>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.OData.Batch.ODataBatchContent.<SerializeToStreamAsync>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.WebHost.HttpControllerHandler.<WriteBufferedResponseContentAsync>d__1b.MoveNext()
```

Looking at the code in SelectExpandNode.cs, it appears the properties to expand are being added to a Dictionary without first checking if the property has already been added to the dictionary. I guess its trying to add twice b/c the AutoExpand has added it, then the $expand option also tries to add the same property.

**Comment by ysmoradi** [comment]
I've the same issue

**Comment by mikepizzo** [comment]
Autoexpand on navigation properties on complex types support released in 7.5.12.
